include/drivers/pinctrl.h

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * Public APIs for pin control drivers
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_PINCTRL_H_
+#define ZEPHYR_INCLUDE_DRIVERS_PINCTRL_H_
+
+/**
+ * @brief Pin Controller Interface
+ * @defgroup pinctrl_interface Pin Controller Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <devicetree.h>
+#include <sys/util.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct device;
+
+/** Pin controller configuration for a given device. */
+struct pinctrl_config {
+	/** Device address. */
+	uintptr_t reg;
+	/** List of states. */
+	const struct pinctrl_state **states;
```

Comment:
I think we should merge the approach here and the one in PR #37621.

such that `struct pinctrl_state` is defined here and the SoC only needs to define `pinctrl_soc_pins_t`

So we'd define the following in here:

```
struct pinctrl_state {
	/** State identifier. */
	enum pinctrl_state_id id;
	/** Pin configurations. */
	const pinctrl_soc_pins_t *pincfgs;
	/** Number of pin configurations. */
	size_t pincfgs_cnt;
};
```

Reply:
Also, do we need `id` or can we index pinctrl_state by id?

Reply:
Agreed, `pinctrl_state` can be generalized as you propose.

Reply:
> Also, do we need `id` or can we index pinctrl_state by id?

I'm not sure I understand, you mean using id as an index to access states?

Reply:
> I'm not sure I understand, you mean using id as an index to access states?

Yes, have pincfgs[] indexed by `enum pinctrl_state_id`.